File: src/services/attendance.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Create a new attendance record.
 * @param attendance - The attendance details.
 * @returns The new attendance ID.
 */
export const createAttendance = async (attendance: {
    staffId: number;
    date: string; // Format: YYYY-MM-DD
    status: "Present" | "Absent";
    reason?: string;
    halfDay: boolean;
    comments?: string;
}): Promise<Object> => {
    try {
        const attendanceId: Object = await invoke("add_attendance", attendance);
        return attendanceId;
    } catch (error) {
        console.error("Error creating attendance:", error);
        throw error;
    }
};

/**
 * Fetch all attendance records.
 * @returns A list of all attendance records.
 */
export const getAllAttendance = async (): Promise<any[]> => {
    try {
        const records = await invoke("get_all_attendance");
        return records as any[] || [];
    } catch (error) {
        console.error("Error fetching attendance records:", error);
        return [];
    }
};

/**
 * Fetch attendance by ID.
 * @param attendanceId - The ID of the attendance record.
 * @returns The attendance record or an empty object if not found.
 */
export const getAttendanceById = async (attendanceId: number): Promise<any> => {
    try {
        const record = await invoke("get_attendance", { attendance_id: attendanceId });
        return record || {};
    } catch (error) {
        console.error("Error fetching attendance record:", error);
        return {};
    }
};

/**
 * Fetch all attendance records for a specific staff member.
 * @param staffId - The ID of the staff member.
 * @returns A list of attendance records.
 */
export const getAttendanceByStaff = async (staffId: number): Promise<any[]> => {
    try {
        const records: any[] = await invoke("get_attendance_by_staff", { staff_id: staffId });
        return records || [];
    } catch (error) {
        console.error("Error fetching staff attendance records:", error);
        return [];
    }
};

/**
 * Update an existing attendance record.
 * @param attendanceId - The ID of the attendance record.
 * @param updates - Updated attendance data.
 * @returns A success message.
 */
export const updateAttendance = async (
    attendanceId: number,
    updates: {
        status: "Present" | "Absent";
        reason?: string;
        halfDay: boolean;
        comments?: string;
    }
): Promise<string> => {
    try {
        const response: string = await invoke("update_attendance", {
            attendance_id: attendanceId,
            ...updates,
        });
        console.log("Attendance Updated:", response);
        return response;
    } catch (error) {
        console.error("Error updating attendance:", error);
        throw error;
    }
};

/**
 * Delete an attendance record.
 * @param attendanceId - The ID of the record to delete.
 * @returns A success message.
 */
export const deleteAttendance = async (attendanceId: number): Promise<string> => {
    try {
        const response: string = await invoke("delete_attendance", { attendance_id: attendanceId });
        console.log("Attendance Deleted:", response);
        return response;
    } catch (error) {
        console.error("Error deleting attendance:", error);
        throw error;
    }
};

File: src-tauri/src/methods/attendance.rs
use rusqlite::params;
use rusqlite::OptionalExtension;
use serde::{Deserialize, Serialize};
use serde_json::json;
use tauri::State;

use crate::AppState; // Import AppState

#[derive(Serialize, Deserialize)]
pub struct Attendance {
    pub attendance_id: i32,
    pub staff_id: i32,
    pub date: String,   // Use ISO-8601 format (YYYY-MM-DD)
    pub status: String, // "Present" or "Absent"
    pub reason: Option<String>,
    pub half_day: bool,
    pub comments: Option<String>,
}

#[tauri::command]
pub fn get_all_attendance(state: State<AppState>) -> Result<Vec<Attendance>, String> {
    let conn = state.conn.lock().unwrap();
    let mut stmt = conn
        .prepare(
            "SELECT attendance_id, staff_id, date, status, reason, half_day, comments FROM attendance",
        )
        .map_err(|e| format!("Failed to prepare statement: {}", e))?;

    let attendance_iter = stmt
        .query_map([], |row| {
            Ok(Attendance {
                attendance_id: row.get(0)?,
                staff_id: row.get(1)?,
                date: row.get(2)?,
                status: row.get(3)?,
                reason: row.get(4)?,
                half_day: row.get(5)?,
                comments: row.get(6)?,
            })
        })
        .map_err(|e| format!("Failed to fetch attendance records: {}", e))?;

    let attendance: Vec<Attendance> = attendance_iter.filter_map(Result::ok).collect();
    Ok(attendance)
}

#[tauri::command]
pub fn add_attendance(
    state: State<AppState>,
    staff_id: i32,
    date: String,
    status: String,
    reason: Option<String>,
    half_day: bool,
    comments: Option<String>,
) -> Result<serde_json::Value, String> {
    let conn = state.conn.lock().unwrap();

    // Validate `status`
    if status != "Present" && status != "Absent" {
        return Err("Invalid status. Must be 'Present' or 'Absent'.".to_string());
    }

    match conn.execute(
        "INSERT INTO attendance (staff_id, date, status, reason, half_day, comments) VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
        params![staff_id, date, status, reason, half_day, comments],
    ) {
        Ok(_) => {
            let new_attendance_id = conn.last_insert_rowid(); // Get last inserted row ID
            Ok(json!({
                "message": "Attendance record added successfully",
                "status": "success",
                "attendance_id": new_attendance_id
            }))
        }
        Err(e) => Err(format!("Failed to add attendance record: {}", e)),
    }
}

#[tauri::command]
pub fn get_attendance(
    state: State<AppState>,
    attendance_id: i32,
) -> Result<Option<Attendance>, String> {
    let conn = state.conn.lock().unwrap();
    let mut stmt = conn
        .prepare(
            "SELECT attendance_id, staff_id, date, status, reason, half_day, comments FROM attendance WHERE attendance_id = ?1",
        )
        .map_err(|e| format!("Failed to prepare statement: {}", e))?;

    let attendance = stmt
        .query_row(params![attendance_id], |row| {
            Ok(Attendance {
                attendance_id: row.get(0)?,
                staff_id: row.get(1)?,
                date: row.get(2)?,
                status: row.get(3)?,
                reason: row.get(4)?,
                half_day: row.get(5)?,
                comments: row.get(6)?,
            })
        })
        .optional()
        .map_err(|e| format!("Failed to fetch attendance record: {}", e))?;

    Ok(attendance)
}

#[tauri::command]
pub fn update_attendance(
    state: State<AppState>,
    attendance_id: i32,
    status: String,
    reason: Option<String>,
    half_day: bool,
    comments: Option<String>,
) -> Result<String, String> {
    let conn = state.conn.lock().unwrap();

    // Validate `status`
    if status != "Present" && status != "Absent" {
        return Err("Invalid status. Must be 'Present' or 'Absent'.".to_string());
    }

    match conn.execute(
        "UPDATE attendance SET status = ?1, reason = ?2, half_day = ?3, comments = ?4 WHERE attendance_id = ?5",
        params![status, reason, half_day, comments, attendance_id],
    ) {
        Ok(_) => Ok("Attendance record updated successfully".to_string()),
        Err(e) => Err(format!("Failed to update attendance record: {}", e)),
    }
}

#[tauri::command]
pub fn delete_attendance(state: State<AppState>, attendance_id: i32) -> Result<String, String> {
    let conn = state.conn.lock().unwrap();
    match conn.execute(
        "DELETE FROM attendance WHERE attendance_id = ?1",
        params![attendance_id],
    ) {
        Ok(_) => Ok("Attendance record deleted successfully".to_string()),
        Err(e) => Err(format!("Failed to delete attendance record: {}", e)),
    }
}

#[tauri::command]
pub fn get_attendance_by_staff(
    state: State<AppState>,
    staff_id: i32,
) -> Result<Vec<Attendance>, String> {
    let conn = state.conn.lock().unwrap();
    let mut stmt = conn
        .prepare(
            "SELECT attendance_id, staff_id, date, status, reason, half_day, comments FROM attendance WHERE staff_id = ?1",
        )
        .map_err(|e| format!("Failed to prepare statement: {}", e))?;

    let attendance_iter = stmt
        .query_map(params![staff_id], |row| {
            Ok(Attendance {
                attendance_id: row.get(0)?,
                staff_id: row.get(1)?,
                date: row.get(2)?,
                status: row.get(3)?,
                reason: row.get(4)?,
                half_day: row.get(5)?,
                comments: row.get(6)?,
            })
        })
        .map_err(|e| format!("Failed to fetch attendance records: {}", e))?;

    let attendance: Vec<Attendance> = attendance_iter.filter_map(Result::ok).collect();
    Ok(attendance)
}
